mlir/lib/Gradient/Transforms/GradMethods/JVPVJPPatterns.cpp

```diff
@@ -0,0 +1,335 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "jvpvjp"
+
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Index/IR/IndexDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Gradient/IR/GradientOps.h"
+#include "Gradient/Utils/CompDiffArgIndices.h"
+#include "Gradient/Utils/EinsumLinalgGeneric.h"
+#include "Gradient/Utils/GradientShape.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/PatternMatch.h"
+
+#include "Gradient/IR/GradientOps.h"
+#include "Gradient/Transforms/Passes.h"
+#include "Gradient/Transforms/Patterns.h"
+#include "Quantum/IR/QuantumOps.h"
+
+#include "JVPVJPPatterns.hpp"
+
+using namespace mlir;
+using namespace catalyst::gradient;
+using llvm::dbgs;
+
+namespace llvm {
+
+template <class T> raw_ostream &operator<<(raw_ostream &oss, const std::vector<T> &v)
+{
+    oss << "[";
+    bool first = true;
+    for (auto i : v) {
+        oss << (first ? "" : ", ") << i;
+        first = false;
+    }
+    oss << "]";
+    return oss;
+}
+}; // namespace llvm
+
+namespace catalyst {
+namespace gradient {
+
+template <class T> std::vector<int64_t> _tovec(const T &x)
+{
+    return std::vector<int64_t>(x.begin(), x.end());
+};
+
+LogicalResult JVPLoweringPattern::matchAndRewrite(JVPOp op, PatternRewriter &rewriter) const
+{
+    MLIRContext *ctx = getContext();
+
+    Location loc = op.getLoc();
+    LLVM_DEBUG(dbgs() << "replacing JVP op\n");
```

Comment:
Are you intending to keep these statements as a logging mechanism? :)

Reply:
Yeah good idea! I hadn't thought about using this feature yet but it is indeed very helpful. Would they print when you specify `--debug` option in `mlir-opt`?